plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+        TermBind a s var rhs ->
+            let rhs' = go rhs
+             in TermBind (a, termUniqs rhs') s (noUniq var) rhs'
+        other -> noUniq other
+
+termUniqs :: Term tyname name uni fun (a, Uniques) -> Uniques
+termUniqs = snd . termAnn
+
+-- | The result includes both the `Unique` of the bound var and the `Unique`s in the RHS.
+bindingUniqs ::
+    PLC.HasUnique name PLC.TermUnique =>
+    Binding tyname name uni fun (a, Uniques) ->
+    Uniques
+bindingUniqs b = Set.union (snd (bindingAnn b)) $ case b of
+    TermBind _a _s var _rhs -> Set.singleton $ var ^. PLC.varDeclName . PLC.unique
+    _                       -> mempty
+
+noUniq :: Functor f => f a -> f (a, Uniques)
+noUniq = fmap (,mempty)
+
+{- | The algebra for floating a `Binding` inwards. Given a `Term` and a `Binding`, determine
+ whether the `Binding` can be placed somewhere inside the `Term`.
+
+ If yes, return the result `Term`. Otherwise, return a `Let` constructed from
+ the given `Binding` and `Term`.
+-}
+floatAlg ::
+    forall tyname name uni fun a.
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.ToBuiltinMeaning uni fun
+    ) =>
+    PLC.BuiltinVersion fun ->
+    -- | Annotation to be attached to the constructed `Let`.
+    a ->
+    Binding tyname name uni fun (a, Uniques) ->
+    Term tyname name uni fun (a, Uniques) ->
+    Term tyname name uni fun (a, Uniques)
+floatAlg ver letAnn = go
+  where
+    go ::
+        Binding tyname name uni fun (a, Uniques) ->
+        Term tyname name uni fun (a, Uniques) ->
+        Term tyname name uni fun (a, Uniques)
+    go b !body = case b of
+        TermBind _a Strict _var rhs
+            -- A strict binding whose RHS is not pure cannot be moved.
+            | not (isPure ver (const NonStrict) rhs) ->
+                let us = Set.union (termUniqs body) (bindingUniqs b)
+                 in Let (letAnn, us) NonRec (pure b) body
+        TermBind (_, usBind) _strictness var _rhs -> case body of
+            Apply (a, usBody) fun arg
+                | (var ^. PLC.unique) `Set.notMember` termUniqs fun ->
+                    -- `fun` does not mention `var`, so we can place the binding inside `arg`.
+                    Apply (a, Set.union usBind usBody) fun (go b arg)
+                | (var ^. PLC.unique) `Set.notMember` termUniqs arg ->
+                    -- `arg` does not mention `var`, so we can place the binding inside `fun`.
+                    Apply (a, Set.union usBind usBody) (go b fun) arg
+            LamAbs (a, usBody) n ty lamAbsBody ->
+                -- A binding can always be placed inside a `LamAbs`.
+                LamAbs (a, Set.union usBind usBody) n ty (go b lamAbsBody)
+            TyAbs (a, usBody) n k tyAbsBody ->
+                -- A binding can always be placed inside a `TyAbs`.
+                TyAbs (a, Set.union usBind usBody) n k (go b tyAbsBody)
+            TyInst (a, usBody) tyInstBody ty ->
+                -- A binding can always be placed inside a `TyInst`.
```

Comment:
```suggestion
                -- A term binding can always be placed inside the body a `TyInst` because the
                -- type cannot mention the bound variable 
```